components/script/dom/mutationobserver.rs

```diff
@@ -37,4 +73,158 @@ impl MutationObserver {
+    /// Notify Mutation Observers.
+    pub fn notifyMutationObservers() -> Fallible<()> {
+        // Step 1
+        ScriptThread::set_mutation_observer_compound_microtask_queued(false);
+        // Step 2
+        let notifyList = ScriptThread::get_mutation_observer();
+        // Step 3, Step 4 not needed as Servo doesn't implement anything related to slots yet.
+        // Step 5: Ignore the specific text about execute a compound microtask.
+        for (i, mo) in notifyList.iter().enumerate() {
+            let queue: Vec<Root<MutationRecord>> = mo.record_queue.borrow().clone();
+            *mo.record_queue.borrow_mut() = Vec::new();
+            // TODO: Step 5.3 Remove all transient registered observers whose observer is mo.
+            if !queue.is_empty() {
+                let callback: MutationCallback = *mo.callback;
+                callback.Call_(&*mo, queue, &*mo, ExceptionHandling::Report);
+            }
+        }
+        // Step 6 not needed as Servo doesn't implement anything related to slots yet.
+        Ok(())
+    }
+
+    //https://dom.spec.whatwg.org/#queueing-a-mutation-record
+    //Queuing a mutation record
+    pub fn queue_a_mutation_record(target: &Node, attr_type: Mutation) {
+        // Step 1
+        let mut interestedObservers: Vec<Root<MutationObserver>> = vec![];
+        let mut pairedStrings: Vec<DOMString> = vec![];
+        // Step 2
+        let mut nodes: Vec<Root<Node>> = vec![];
+        for ancestor in target.inclusive_ancestors() {
+            nodes.push(ancestor);
+        }
+        // Step 3
+        for node in &nodes {
+            for registered_observer in node.registered_mutation_observers().borrow().iter() {
+                match attr_type {
+                    // TODO check for Mutations other than Attribute
+                    Mutation::Attribute { name, namespace, oldValue, newValue } => {
+                        let condition1: bool = node != &Root::from_ref(target) &&
+                            !registered_observer.subtree.get();
+                        let condition2: bool = registered_observer.attributes.get() == false;
+                        let condition3: bool = !registered_observer.attribute_filter.borrow().is_empty() &&
+                            (registered_observer.attribute_filter.borrow().iter()
+                                .find(|s| **s == DOMString::from(&*name)).is_none() || namespace != ns!());
+                        if !condition1 && !condition2 && !condition3 {
+                            // Step 3.1
+                            if !interestedObservers.contains(registered_observer) {
+                                interestedObservers.push(Root::from_ref(registered_observer));
+                            }
+                            // Step 3.2
+                            if registered_observer.attribute_old_value.get() == true {
+                                pairedStrings.push(DOMString::from(oldValue));
+                            }
```

Comment:
This avoids all of the problems with matching indexes or pushing placeholder values.
```rust
// Step 3.1.2
let paired_string = if registered_observer.attribute_old_value.get() {
    DOMString::from(oldValue)
} else {
    DOMString::from("")
};
// Step 3.1.1
let idx = interestedObservers.iter().position(|&(ref o, _)| &**o as *const MutationObserver == &**registered_observer as *const MutationObserver);
if let Some(idx) = idx {
    interestedObservers[idx].1 = paired_string;
} else {
    interestedObservers.push((Root::from_ref(registered_observer), paired_string));
}
```